src/read.rs

```diff
@@ -1684,18 +1677,221 @@ pub fn read_zipfile_from_stream<'a, R: Read>(reader: &'a mut R) -> ZipResult<Opt
+    /// creates a new instance of this, first reads from the buffer, then proxy read to underlying stream
+    pub fn new(reader: &'a mut T, buffer: Vec<u8>) -> Self {
+        ReadBufferThenForward {
+            reader,
+            buffer,
+            index: 0,
+        }
+    }
+}
+
+impl<'a, T: Read> Read for ReadBufferThenForward<'a, T> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        let buffer_start = std::cmp::min(self.index, buf.len());
+        let data_left_in_buffer = buf.len() - buffer_start;
+        let buffer_end = std::cmp::min(self.buffer.len(), buffer_start + data_left_in_buffer);
+
+        let next_buffer_data = &self.buffer[buffer_start..buffer_end];
+        let count_copy_from_buffer = std::cmp::min(next_buffer_data.len(), buf.len());
+
+        buf[0..count_copy_from_buffer]
+            .copy_from_slice(&next_buffer_data[0..count_copy_from_buffer]);
+
+        self.index += count_copy_from_buffer;
+        if self.index > buf.len() && self.index > 0 {
+            // empty buffer
+            self.index = 0;
+            self.buffer.clear();
+            self.buffer.shrink_to_fit();
+        }
+
+        let remaining_bytes = buf.len() - count_copy_from_buffer;
+        let result = if remaining_bytes > 0 {
+            self.reader.read(&mut buf[count_copy_from_buffer..])?
+        } else {
+            0
+        };
+
+        Ok(count_copy_from_buffer + result)
+    }
+}
+
+struct ReadTillDataDescriptor<T: Read> {
+    reader: T,
+    data_descriptor_found: Option<ZipDataDescriptor>,
+    look_ahead_buffer: Vec<u8>,
+    number_read_total_actual: usize, // without look ahead buffer
+}
+
+impl<T: Read> ReadTillDataDescriptor<T> {
+    pub fn new(mut reader: T) -> ZipResult<Self> {
+        let mut look_ahead_buffer = [0u8; std::mem::size_of::<ZipDataDescriptor>()];
+
+        // fill the look ahead buffer
+        reader.read_exact(&mut look_ahead_buffer)?;
+
+        let look_ahead_buffer = Vec::from(look_ahead_buffer);
+
+        Ok(Self {
+            reader,
+            look_ahead_buffer,
+            data_descriptor_found: None,
+            number_read_total_actual: 0,
+        })
+    }
+}
+
+impl<T: Read> ReadTillDataDescriptor<T> {
+    fn has_found_data_descriptor(&self) -> Option<ZipDataDescriptor> {
+        if spec::Magic::from_first_le_bytes(&self.look_ahead_buffer)
+            == spec::Magic::DATA_DESCRIPTOR_SIGNATURE
+        {
+            // potentially found a data descriptor
+            // check if size matches
+            let data_descriptor = match ZipDataDescriptor::interpret(&self.look_ahead_buffer) {
+                Ok(data_descriptor) => data_descriptor,
+                Err(_) => return None,
+            };
+            let data_descriptor_size = data_descriptor.compressed_size;
+
+            if data_descriptor_size == self.number_read_total_actual as u32 {
+                return Some(data_descriptor);
+            }
+        }
```

Comment:
`interpret` already checks whether the magic matches.

```suggestion
        let Ok(data_descriptor) = ZipDataDescriptor::interpret(&self.look_ahead_buffer) else {
           return None;
        };
        let data_descriptor_size = data_descriptor.compressed_size;

        if data_descriptor_size == self.number_read_total_actual as u32 {
            Some(data_descriptor)
        } else {
            None
        }
```

Reply:
Check the CRC32 here as well.

Reply:
This probably needs some architectural changes. The function above operates on the compressed zip file. The CRC is calculated from the uncompressed data. With the current architecture checking CRC in this function is not possible.

Currently I have no good idea to solve this. Still, the CRC is checked, so if an "data descriptor" occurs in the stream - Which is only the case if the file size in the data descriptor is correct - the CRC is likely wrong. In this case only the first half/or part of the file is returned to the user, but then an error is thrown when the CRC check fails; reaching the stream EOF.